keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -687,20 +997,92 @@ We will extend this approach:
 Note that this does allow `ValidatingAdmissionPolicy` to match
 `ValidatingWebhookConfiguration`.
 
-We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
-configuration CRD kind that is uses for config. It would be possible to recover
-from this since the `ValidatingAdmissionPolicy` would remain accessible, but
-might require a privileged user do the repair.
-
-TODO: We should consider loosening this up and allow admission configuration to
-intercept/guard writes to admission configuration while preventing deadlock -
-Add feature to configure a set of webhooks to intercept other webhooks
-https://github.com/kubernetes/kubernetes/issues/101794.
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
 
 Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
 `ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
 higher level. This could be added later.
 
+#### Reporting violations
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+- Ability to format message strings
+- support for reasons/codes
+
+Example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  ...
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      level: Deny # the default
```

Comment:
Different enforcement points may want different enforcement actions and may have different possible actions they can take. If the goal is to have this syntax shift left/right, we want to tread carefully here.

Another option would be to let policy frameworks handle the higher-level consideration and assume this is purely for admission-time enforcement.

Reply:
I _really_ want to put enforcement action under the control of the policy configurations.

Things that come to mind:
- Should each validation expression in a policy be able to have it's own enforcement action? (or should we just agree that all the validation expressions of policy share enforcement action?)
- What happens two policy definitions share a configuration, seems like the only option is to have the configuration's enforcement action effect both policies. How does one roll out a policy in this model?

Reply:
I've tried moving enforcement to the configurations by requiring names for all validation expressions and then requiring that policy configurations set the level of each validation expression (this is also useful for audit which needs a key for each violation).

Reply:
Moved enforcement control to policy configuration.